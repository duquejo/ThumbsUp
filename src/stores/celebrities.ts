import { computed, reactive, type ComputedRef } from 'vue';
import { defineStore } from 'pinia';
import { data } from '@/data.json';
import type { ICelebrity } from '@/interfaces/celebrities';

export type VoteTypes = 'positive' | 'negative';

export const useCelebritiesStore = defineStore('celebrities', () => {
  const celebrities = reactive<ICelebrity[]>(data);

  const getRandomCelebrity = computed<ICelebrity>(() => {
    const rand = Math.floor(Math.random() * celebrities.length);
    return celebrities[rand];
  });

  return {
    // State
    celebrities, 
    
    // Actions
    celebrityVote: (id: number, vote: VoteTypes) => {
      celebrities.map((celeb: ICelebrity) => {
        if (celeb.id === id) {
          return {
            ...celeb,
            votes: {
              positive: vote === 'positive' ? celeb.votes.positive++ : celeb.votes.positive,
              negative: vote === 'negative' ? celeb.votes.negative++ : celeb.votes.negative,
            },
          };
        }
        return celeb;
      });
    },

    getCelebrityStatusById: (id: number) => {
      const celebrity = celebrities.find((celeb: ICelebrity) => celeb.id === id);
      if (!celebrity) throw new Error('Celebrity not found');
      return celebrity.votes.positive >= celebrity.votes.negative ? 'positive' : 'negative';
    },
    
    // Getters
    getRandomCelebrity
  };
});
