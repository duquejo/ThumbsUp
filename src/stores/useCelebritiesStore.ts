import { ref } from 'vue';
import { defineStore } from 'pinia';
import type { ICelebrity } from '@/interfaces/Celebrities';
import useLocalStorage from '@/composables/useLocalStorage';
import { postCelebrityVote } from '@/api/celebritiesApi';

export type VoteTypes = 'positive' | 'negative';

export const useCelebritiesStore = defineStore('celebrities', () => {
  const { saveItem } = useLocalStorage('celebrities');

  const celebrities = ref<ICelebrity[]>([]);

  return {
    // State
    celebrities,

    // Actions
    celebrityVote: (id: number, vote: VoteTypes) => {
      celebrities.value.map((celeb: ICelebrity) => {
        if (celeb.id === id) {
          return {
            ...celeb,
            votes: {
              positive: vote === 'positive' ? celeb.votes.positive++ : celeb.votes.positive,
              negative: vote === 'negative' ? celeb.votes.negative++ : celeb.votes.negative,
            },
          };
        }
        return celeb;
      });
      if( import.meta.env.VITE_VOTES_URL ) {
        postCelebrityVote(id, vote);
      }
      saveItem(celebrities.value);
    },
    setCelebritiesState: (newCelebs: ICelebrity[]) => {
      celebrities.value = newCelebs;
      saveItem(newCelebs);
    },
  };
});
