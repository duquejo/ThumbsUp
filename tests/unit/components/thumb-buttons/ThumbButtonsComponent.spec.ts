import ThumbButtonsComponentVue from '@/components/thumb-buttons/ThumbButtonsComponent.vue';
import { VueWrapper, shallowMount } from '@vue/test-utils';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { createTestingPinia } from '@pinia/testing';
import { useCelebritiesStore } from '@/stores/useCelebritiesStore';

vi.mock('@/stores/useCelebritiesStore', () => ({
  useCelebritiesStore: vi.fn(() => ({
    celebrityVote: vi.fn()
  }))
}));

const useCelebrityStoreMock = vi.mocked(useCelebritiesStore);

describe('+ ThumbButtonsComponent unit tests', () => {

  const properties = {
    category: 'entertainment',
    lastUpdated: '2020-03-10T23:08:57.892Z',
    id: 1,
  };

  let wrapper: VueWrapper;

  beforeEach(() => {
    vi.clearAllMocks();
    wrapper = shallowMount(ThumbButtonsComponentVue, {
      props: {  ...properties },
      global: {
        plugins: [createTestingPinia({
          createSpy: vi.fn(),
        })],
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('should render properly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should handle the buttons \'idle\' state', () => {
    expect(wrapper.html()).toMatchSnapshot();

    const buttons = wrapper.findAll('button');

    expect(buttons[0].attributes('aria-label')).toBe('thumbs up');
    expect(buttons[1].attributes('aria-label')).toBe('thumbs down');
    expect(buttons[2].attributes('title')).toBe('You must thumb up/down first');
    expect(buttons[2].attributes('disabled')).toBe('');
  });

  it('should handle the buttons \'ready\' state - Positive', async () => {
    const buttons = wrapper.findAll('button');

    await buttons[0].trigger('click');

    expect(wrapper.html()).toMatchSnapshot();

    expect(buttons[0].classes('selected')).toBeTruthy();
    expect(buttons[1].classes('selected')).toBeFalsy();
    expect(buttons[2].attributes('title')).toBe('Click to vote');
    expect(buttons[2].attributes('disabled')).toBeFalsy();
  });

  it('should handle the buttons \'ready\' state - Negative', async () => {
    const buttons = wrapper.findAll('button');

    await buttons[1].trigger('click');

    expect(wrapper.html()).toMatchSnapshot();

    expect(buttons[1].classes('selected')).toBeTruthy();
    expect(buttons[0].classes('selected')).toBeFalsy();
    expect(buttons[2].attributes('title')).toBe('Click to vote');
    expect(buttons[2].attributes('disabled')).toBeFalsy();
  });

  it('should handle the buttons \'done\' state', async () => {
    const buttons = wrapper.findAll('button');

    await buttons[0].trigger('click'); // Changing Idle state
    await buttons[2].trigger('click'); // Vote
    

    const eyebrow = wrapper.find('.eyebrow');
    const postClickButtons = wrapper.findAll('button');

    expect(wrapper.html()).toMatchSnapshot();

    expect(buttons[2].text()).toBe('Vote again');
    expect(eyebrow.text()).toBe('Thank you for voting!');

    expect(postClickButtons).toHaveLength(1);

    expect(useCelebrityStoreMock).toHaveBeenCalledTimes(1);
    expect(useCelebrityStoreMock).toHaveBeenCalledWith();
  });
});