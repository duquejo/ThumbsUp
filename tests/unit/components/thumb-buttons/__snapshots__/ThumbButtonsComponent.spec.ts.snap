// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`+ ThumbButtonsComponent unit tests > should handle the buttons 'done' state 1`] = `
"<!-- Eyebrow data -->
<span data-v-f6c00734="" class="eyebrow">Thank you for voting!</span>
<div data-v-f6c00734="" class="card__thumbs-container">
  <!--v-if-->
  <!--v-if--><button data-v-f6c00734="" type="submit" title="Click to vote">Vote again</button>
</div>"
`;

exports[`+ ThumbButtonsComponent unit tests > should handle the buttons 'idle' state 1`] = `
"<!-- Eyebrow data -->
<eyebrow-component-stub data-v-f6c00734="" lastupdated="2020-03-10T23:08:57.892Z" category="entertainment"></eyebrow-component-stub>
<div data-v-f6c00734="" class="card__thumbs-container"><button data-v-f6c00734="" class="icon-button card-thumb-button" aria-label="thumbs up"><img data-v-f6c00734="" src="/assets/img/thumbs-up.svg" alt="thumbs up"></button><button data-v-f6c00734="" class="icon-button card-thumb-button" aria-label="thumbs down"><img data-v-f6c00734="" src="/assets/img/thumbs-down.svg" alt="thumbs down"></button><button data-v-f6c00734="" type="submit" disabled="" title="You must thumb up/down first">Vote now</button></div>"
`;

exports[`+ ThumbButtonsComponent unit tests > should handle the buttons 'ready' state - Negative 1`] = `
"<!-- Eyebrow data -->
<eyebrow-component-stub data-v-f6c00734="" lastupdated="2020-03-10T23:08:57.892Z" category="entertainment"></eyebrow-component-stub>
<div data-v-f6c00734="" class="card__thumbs-container"><button data-v-f6c00734="" class="icon-button card-thumb-button" aria-label="thumbs up"><img data-v-f6c00734="" src="/assets/img/thumbs-up.svg" alt="thumbs up"></button><button data-v-f6c00734="" class="icon-button card-thumb-button selected" aria-label="thumbs down"><img data-v-f6c00734="" src="/assets/img/thumbs-down.svg" alt="thumbs down"></button><button data-v-f6c00734="" type="submit" title="Click to vote">Vote now</button></div>"
`;

exports[`+ ThumbButtonsComponent unit tests > should handle the buttons 'ready' state - Positive 1`] = `
"<!-- Eyebrow data -->
<eyebrow-component-stub data-v-f6c00734="" lastupdated="2020-03-10T23:08:57.892Z" category="entertainment"></eyebrow-component-stub>
<div data-v-f6c00734="" class="card__thumbs-container"><button data-v-f6c00734="" class="icon-button card-thumb-button selected" aria-label="thumbs up"><img data-v-f6c00734="" src="/assets/img/thumbs-up.svg" alt="thumbs up"></button><button data-v-f6c00734="" class="icon-button card-thumb-button" aria-label="thumbs down"><img data-v-f6c00734="" src="/assets/img/thumbs-down.svg" alt="thumbs down"></button><button data-v-f6c00734="" type="submit" title="Click to vote">Vote now</button></div>"
`;

exports[`+ ThumbButtonsComponent unit tests > should render properly 1`] = `
"<!-- Eyebrow data -->
<eyebrow-component-stub data-v-f6c00734="" lastupdated="2020-03-10T23:08:57.892Z" category="entertainment"></eyebrow-component-stub>
<div data-v-f6c00734="" class="card__thumbs-container"><button data-v-f6c00734="" class="icon-button card-thumb-button" aria-label="thumbs up"><img data-v-f6c00734="" src="/assets/img/thumbs-up.svg" alt="thumbs up"></button><button data-v-f6c00734="" class="icon-button card-thumb-button" aria-label="thumbs down"><img data-v-f6c00734="" src="/assets/img/thumbs-down.svg" alt="thumbs down"></button><button data-v-f6c00734="" type="submit" disabled="" title="You must thumb up/down first">Vote now</button></div>"
`;
